### DISCUSSION TITLE
Support inserting a single proto.block

### DISCUSSION THREAD

**Discussion by zdyj3170101136** [ClickHouse/clickhouse-go#1408] [Feature requests]
Currently, an insert is implemented like this.
1. Prepare batch insert
2. Append columns into batch.
3. Send block to clickhouse.

```
tx, err := conn.PrepareBatch()
if err != nil {
return nil, err
}
tx.Append()
tx.Send()
```
Usually, inserts are aggregated in the backend and inserted once every 100,000 records, which leads to very high memory usage of the service.

The block and user's []slice structures need to be released only after all data is successfully inserted, this would cost 10~ seconds.

If block-level insertion is allowed, the user's []slice structure can be released as soon as possible to reduce memory usage.

1, convert []slice to block
2, release []slice
3, send block to backend.

**Comment by zdyj3170101136** [comment]
@jkaflik cc

**Comment by jkaflik** [comment]
@zdyj3170101136 this sounds you are more interested with API like https://github.com/ClickHouse/ch-go does provide. What do you think?

**Reply by zdyj3170101136** [reply]
ch-go has some problems:
First, when it adds col.Str to proto.block, col.Str is not released, resulting in high memory usage.
Second, the buf of proto.block is not allocated in advance, which may lead to excessive memory usage.

<img width="1101" alt="截屏2024-09-18 上午11 46 32" src="https://github.com/user-attachments/assets/3decff9b-6ce4-4432-8fdf-000e3fa69ef0">

**Reply by zdyj3170101136** [reply]
i suggest the proto.block is net.Buffers.

the net.Buffers could use api WriteTo send multi [][]byte to conn.

**Reply by zdyj3170101136** [reply]
> i suggest the proto.block is net.Buffers.
> 
> the net.Buffers could use api WriteTo send multi [][]byte to conn.

i start a pr for this https://github.com/ClickHouse/ch-go/pull/421.
